### ISSUE TITLE
2.4.7/2.4.11/2.4.12 Focus: should indicator be unique?

### ISSUE THREAD

**Issue by Jym77** [w3c/wcag#1383]
Nothing in these SC prevent several components to share the exact same focus indicator.
I read the Understanding document of 2.4.7 as this being the intention (user needs to know where they are, which is hard is focus indicator is the same).

But while trying to write an ACT rule for 2.4.7 we ran into discussion about this precise point and concluded that 2.4.7 does not require unicity of the indicators: https://github.com/act-rules/act-rules.github.io/pull/1329#discussion_r469826197 and following discussion.

If the intention is to have unique indicators, this should be mentioned in the SC. Maybe we can't change 2.4.7 but it is worth adding to 2.4.11/2.4.12.

**Comment by alastc** [comment]
Hi @Jym77,

I'm really not sure what you mean by 'unique' in this context?

For example, if every button/link/control has a blue outline (on a white background), do you mean those are 'identical'?

**Comment by patrickhlauke** [comment]
i was also having trouble trying to work this out, but i *think* - based on one of @WilcoFiers' examples of the hidden video buttons - it means that nothing prohibits controls from appearing in the same place as an another control, and then having the same/similar focus control, visually (but only because the newly focused control that was previously hidden now occupies the space of another control that is now hidden) ... or something? i.e. a very edge-casey situation, and making sure that WCAG does not disqualify this. any other scenario would be rather odd, unless i'm missing something

**Comment by patrickhlauke** [comment]
if there are other scenarios that you have in mind, @Jym77 ... maybe a mockup/picture could help here...

**Comment by JAWS-test** [comment]
If I have understood it correctly, the point is that no SC requires that the focus is located at the focused element and can be clearly assigned to it. I.e. it could be:
* menu items in a menu bar get the focus and only the menu bar is shown as focused because the focus indicator of each menu item extends over the whole menu
* the focus indicator is only to the left of a menu item and it is not apparent whether it refers to the menu item to the left or right of it

**Comment by patrickhlauke** [comment]
that would be a very bizarre interpretation (even for me, famous for hair-splitting). normatively the SC talks about "the keyboard focus indicator is visible", i.e. the indicator of where keyboard focus is is visible (though we had discussions about 'keyboard focus' meaning 'the thing that has the current regard/would be activated if you pressed the activation key/etc', not 'what element is `document.activeElement`). "*the*" indicator, not "an" indicator. so it definitely wants "the indicator" to be on what has focus.

**Comment by JAWS-test** [comment]
@patrickhlauke In my opinion, the 4 SCs for focus, which we will have in WCAG 2.2, does not define anywhere that the focus should be at the element and must be clearly assignable to the element. 

The indicator may or may not be unique, but how do I know which indicator belongs to which element? My [first example](https://codepen.io/jaws-test/pen/KKzyrXO) is indeed bizarre (although it probably does not violate any WCAG-SC), but my [second example](https://codepen.io/jaws-test/pen/xxxJNQY) is by no means bizarre, because I have encountered it many times in WCAG testing. I always judged it as a violation of 2.4.7, but that was my subjective interpretation, which does not do justice to the wording of 2.4.7.

**Comment by JAWS-test** [comment]
I created [two more examples](https://codepen.io/jaws-test/pen/qBZpWdw) where the focus is not assignable and probably not prohibited by SCs:
* focus indicator is displayed for multiple items
* several elements share one focus indicator

I already had the second example during a WCAG test. The problem occurred in a tree structure. Each tree item had several elements (arrows to show and hide, text to open, checkbox to select). For all three elements (which individually got the focus with the tab key; it was not a tree with ARIA roles), the focus indicator was only displayed for the entire tree item and not for each element within the tree item

**Comment by awkawk** [comment]
I would say that both of those examples fail, as they don't meet the common understanding of what focus means.

We didn't define focus in WCAG 2.0, because we felt that the generally accepted definition of focus was sufficient. The focus is generally understood to help users interact with the UI by highlighting which item is able to be interacted with. Neither of your examples clarify this for the user.

So, I agree with the original question - focus indicators _should_ be unique, but there are edge cases that may be sufficient also. I do not think that we should try to lock this down by defining focus. Doing so would likely create more problems than it solves.

The 2.4.7 understanding document uses this common understanding when it starts out:

> The purpose of this success criterion is to help a person know which element has the keyboard focus.
> 
> It must be possible for a person to know which element among multiple elements has the keyboard focus.

This text is non-normative, but I believe that the intent of the SC for focus are already clear, even if it isn't possible to define an automatable rule to catch all issues.

**Comment by Jym77** [comment]
The case I was thinking about was the second example in this @JAWS-test codepen: https://codepen.io/jaws-test/pen/qBZpWdw
"Several element share one focus indicator".

While writing the ACT rule, the first iteration had a requirement that focus indicators are not shared, but then the discussion ensued and we concluded that WCAG does not specifically requires this.
Even if ACT rules aims at automation, there are also manual test rules, and the first goal is consistency of the results among testers.


A case I can imagine where sharing a focus indicator is maybe OK could be, say, coloured buttons and the focus indicator for same is the same `div` but in a colour matching the one of the focused button (so indicator is blue if "blue button" is focused and red if "red button" is focused). This might have problems with "Use of colors", but probably not for 2.4.7…
OTOH, this is also not really "shared" indicator. They have different indicators which happen to be in the same place (they differ by colour).


Back to the question, I think that something like @JAWS-test example of "shared focus indicator" is failing 2.4.7. Would there be cases where something like that actually passes 2.4.7? Should we strengthen it (or 2.4.11) to require "no shared indicator"?
In the current state of the SC, we can't easily fail that case in an ACT rule. This is sort of OK because ACT rules are meant to flag clear failures and often when an ACT rule passes, further tests are needed to ensure that the SC is satisfied. But if "no shared" becomes a clear requirement in a SC, then we can have a rule that flags it for sure. (whether these rules are fully automated or manual is another question)

**Comment by patrickhlauke** [comment]
> "Several element share one focus indicator"

again, to me that would go against the idea of "keyboard focus indicator", as the keyboard focus is not indicating me where the keyboard focus is (on link 1, link 2, link 3, link 4?) but just handwaving to some general "it's somewhere here" area, which is pointless ... and I personally don't think the normative wording could be misread/misunderstood to imply that that's ok.

> A case I can imagine where sharing a focus indicator is maybe OK could be

that's quite an unusual/obtuse idea, but I "guess" one could argue that that does give an unambiguous indication of which element has focus (though as you note, it may then fail other SCs in the process)

>  Should we strengthen it (or 2.4.11) to require "no shared indicator"

I think the whole concept/idea of having a "shared indicator" is so unusual/alien/counterintuitive (maybe just me, but I didn't understand what you were going on about originally) because of the very fact that such a shared indicator ... fails to actually indicate what does and doesn't currently have keyboard focus, that personally I don't think it needs clarification. I mean the clarification would be "when we say keyboard focus indicator, we mean something that makes it clear visually which specific element currently has focus" which seems more like a truism / tautology.

**Comment by Jym77** [comment]
@patrickhlauke I essentially agree with your comment (hence the "unicity" requirement I initially put in the ACT rule).
Discussion with @WilcoFiers lead to removing that requirement as he is convinced that "shared focus indicator" is currently allowed by WCAG…
You're more than welcome to chime in the ACT rule discussion I link earlier and add your opinion, of course.

Part of the question is whether WCAG considers that the edge cases where "shared focus indicator" is OK are rare enough that it is worth completely banning them. Or whether we consider that it is a frequent enough pattern that we need to accept it somehow…
A possible answer might be that 2.4.7 (which is quite weak anyway) accepts "shared focus indicator" but 2.4.11/2.4.12 don't (hence this thread 😁)

**Comment by patrickhlauke** [comment]
i got the impression - at least from @WilcoFiers' example of two video play button things, if i understood it right - that he's not necessarily disagreeing that the focus indicator needs to be unambiguous, but rather that it's not a simple mechanistic "doesn't occupy the same pixels as another elements' focus indicator" type rule, exactly because that rule can't take account of things like one control disappearing, another taking its exact place, and therefore focus indicators essentially overlapping/being the same in terms of pixels on the screen. but maybe i've missed something in my very high-level skim read of that thread at the time.

**Comment by patrickhlauke** [comment]
or, more broadly, i think making some kind of rule that just looks at pixels works, as that's a far too mechanistic and reductive rule that doesn't take into account that things may move/change/appear/disappear leading to focus indicators nominally occupying the same pixels, but actually referring to a different element when the visibility of elements etc changes. unless you then start qualifying that rule/test about pixels with LOTS of other sub-rules ("unless the elements occupy the same on-screen space, but are not visible concurrently" etc), which starts getting very obtuse to understand.

**Comment by mraccess77** [comment]
While the indicator could be in a shared location - it's my opinion that 2.4.7 requires it be clear (unambiguous) which control receives keyboard operation when navigating.  As an aside, there can be times when focus is lost or when keyboard shortcuts work for non-focused controls, or default keystrokes for forms work to submit a form, etc.  but I don't think this test is aimed at these other situations such as default buttons, etc.

**Comment by JAWS-test** [comment]
I think that in most cases it is clear whether a focus indicator belongs to an element or not - making it easy to decide whether the SCs for focus are met. But I still consider the example at https://codepen.io/jaws-test/pen/xxxJNQY to be a difficult borderline case, because it is not possible to tell to which element the indicator belongs. It can only be deduced from the context when I navigate through the application and remember where the focus was displayed at the first element. I have had such examples several times in WCAG tests and evaluated them as a violation of SC 2.4.7. My question would be, do other testers see it the same way?
And perhaps it should be mentioned in the Understanding that a focus indicator must be clearly assignable to an element.

**Comment by patrickhlauke** [comment]
personally i'd say that example nominally passes 2.4.7, but would make a best practice recommendation that it's not overly clear. not sure if adding an explicit wording of "must be clearly assignable visually" or similar would help, as it would still need a subjective assessment...

**Comment by alastc** [comment]
Draft response:
<hr>
Focus indicators should be unique, but there are edge cases that may be sufficient also. Therefore it is not appropriate to lock this down by defining focus. Doing so would likely create more problems than it solves.

The ACT rule discussion appears to be talking about controls that appear in the same place, rather than focus indicators that are confusingly placed.

**Comment by alastc** [comment]
Hi @Jym77,

The group agreed this response today:
<hr>
Focus indicators should be unique and it is best practice to do make them so, but there are edge cases that may be sufficient also. The intent of this SC is not to force focus indicators to be unique by defining focus. Doing so would likely create more problems than it solves but the issue can be explored further in a future version of WCAG.
<hr>
